enhancements/network/optional-on-prem-dns-records.md

```diff
@@ -0,0 +1,219 @@
+---
+title: optional-on-prem-dns-records
+authors:
+  - bnemec
+reviewers:
+  - mko # On-Prem Networking
+  - everettraven # API
+approvers:
+  - knobunc
+api-approvers:
+  - JoelSpeed
+creation-date: 2025-05-21
+last-updated: 2025-05-21
+tracking-link:
+  - https://issues.redhat.com/browse/OPNET-678
+see-also:
+  - "/enhancements/network/external-lb-vips.md"
+replaces:
+  - NA
+superseded-by:
+  - NA
+---
+
+# Optional On-Prem DNS Records
+
+## Summary
+
+Allow deployers to disable the internal DNS records for api, api-int, and
+ingress when external loadbalancers are used.
+
+## Motivation
+
+Previously when an external loadbalancer was configured we still used the VIP
+fields of install-config to create DNS records in the internal DNS
+infrastructure pointing at the external loadbalancer. However, some more
+advanced loadbalancers also use DNS loadbalancing to further distribute
+traffic. This does not work if we hard-code a single loadbalancer endpoint
+into our internal DNS.
+
+### User Stories
+
+* As an OpenShift deployer I want to use an advanced external loadbalancer
+  to distribute traffic to my cluster. This loadbalancer uses multiple DNS
+  targets in order to reduce load on a single endpoint.
+
+### Goals
+
+Make it possible to disable the internal DNS records for api, api-int, and
+ingress in on-prem IPI deployments to allow use of external DNS to manage
+those, even for internal cluster traffic.
+
+### Non-Goals
+
+Disabling the internal DNS infrastructure entirely. There are two reasons
+we're not doing this:
+
+* It also provides records for each host in the cluster, although it's not
+  clear this is actually necessary anymore.
+* It would require significant re-working of the resolv-prepender behavior,
+  and that component has historically been rather fragile. Adding complexity
+  in that area is more likely to cause problems and would run counter to past
+  efforts to simplify it.
+
+## Proposal
+
+Add a configuration option for the internal DNS records so they can be disabled
+when an external loadbalancer is in use. When this configuration is used, the
+deployer will be wholly responsible for providing loadbalancer and DNS services
+to the cluster.
+
+Note that api and ingress are already required external records, so in reality
+the only additional requirement here is api-int.
+
+### Workflow Description
+
+The deployer will configure external DNS and loadbalancing, similar to how
+they would for a UPI deployment. They will then deploy an IPI cluster with
+the loadbalancer set to UserManaged and DNS set to disabled.
+
+### API Extensions
+
+A new field will be added to control whether DNS records are deployed.
```

Comment:
Presumably this is the `internalDNSRecords` field mentioned further down in the document?

Would you mind adding a snippet of the proposed Go changes here? I'd also accept a link to a PR to openshift/api where we can go through a proper API review on the proposed changes.

Reply:
Looks like that API change is only a portion of the changes, presumably because the installer sets the status there.

Could you also include the proposed installer `InstallConfig` API changes and validations here?

Reply:
Sure, that's here: https://github.com/openshift/installer/pull/9924

Reply:
Thanks!

I left some comments on the openshift/api PR with some thoughts.

Once we've pinpointed the API changes that need to take place, including links to the PRs where we've had discussions or explicitly including the code in the EP would be good.